### ISSUE TITLE
A list of widgets breaking strict Content-Security-Policy (CSP) directives

### ISSUE THREAD

**Issue by brownaa** [wagtail/wagtail#7053]
## Is your proposal related to a problem?

See #1288. Wagtail uses inline javascript and styles to provide useful feedback to the user. This causes conflicts with CSP directives.

This issue attempts to list all of the potential places this occurs in the Wagtail codebase.

## Directives to be reviewed:

- [x] script-src ((completed 2021-04-21))
- [x] img-src ((completed 2024-09-27))
- [x] ~font-src~ (no longer relevant)
- [x] style-src ((completed 2021-04-21))

## Extra Notes

- The JS library `modernizer-2.6.2.min.js` is blocked by the `style-src: 'self'` CSP. You can trigger this by visiting /admin/

## img-src

```
img-src 'self' data: blob:;
```

See #12368.

## font-src 

No Wagtail font files as of a few years ago.

### Wagtail Admin

- [ ] [wagtail/admin/edit_handlers.py](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/edit_handlers.py#L30)

#### Wagtail Admin Templates
1. wagtail/admin/templates/wagtailadmin/ 

    - [ ] [base.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/admin_base.html#L18-L41)
    - [ ] [home.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/home.html#L36-L56)
    - [x] [login.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/login.html#L97-L101)
    - [ ] [skeleton.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/skeleton.html#L23-L70)

1. wagtail/admin/templates/wagtailadmin/edit_handlers/

    - [ ] [wagtail/admin/templates/wagtailadmin/edit_handlers/inline_panel.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/edit_handlers/inline_panel.html#L20-L24)
    need to inspect how code is injected here to see if there are more ways this might break CSP

1. wagtail/admin/templates/wagtailadmin/home/

    - [ ] [wagtail/admin/templates/wagtailadmin/home/locked_pages.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/home/locked_pages.html#L54-L56)
    - [ ] [wagtail/admin/templates/wagtailadmin/home/workflow_pages_to_moderate.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/home/workflow_pages_to_moderate.html#L80-L93)

1. wagtail/admin/templates/wagtailadmin/pages/

    - [ ] [wagtail/admin/templates/wagtailadmin/pages/listing/_button_with_dropdown.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/pages/listing/_button_with_dropdown.html#L3)
    NOTE: need to inspect if this breaks CSP
    - [ ] [wagtail/admin/templates/wagtailadmin/pages/workflow_history/detail.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/pages/workflow_history/detail.html#L144-L157)
    - [ ] [wagtail/admin/templates/wagtailadmin/pages/workflow_history/index.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/pages/workflow_history/index.html#L25-L36)
    - [ ] [wagtail/admin/templates/wagtailadmin/pages/_editor_js.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/pages/_editor_js.html#L8-L17)
    - [ ] [wagtail/admin/templates/wagtailadmin/pages/create.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/pages/create.html#L100-L117)
    NOTE: see comments above code linked here and check if relevant of if CSP still interprets this as an inline script
    - [ ] [wagtail/admin/templates/wagtailadmin/pages/edit.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/pages/edit.html#L117-L182)
    NOTE: see comments above code linked here and check if relevant of if CSP still interprets this as an inline script
    - [ ] [wagtail/admin/templates/wagtailadmin/pages/index.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/pages/index.html#L33-L82)
    - [ ] [wagtail/admin/templates/wagtailadmin/pages/preview_error.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/pages/preview_error.html#L18)
    NOTE: need to inspect if this breaks CSP
    - [ ] [wagtail/admin/templates/wagtailadmin/pages/search.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/pages/search.html#L7-L13)

1. wagtail/admin/templates/wagtailadmin/permissions/includes/

    - [ ] [wagtail/admin/templates/wagtailadmin/permissions/includes/collection_member_permissions_formset.html#L45-L51](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/permissions/includes/collection_member_permissions_formset.html#L45-L51)
    - [ ] [wagtail/admin/templates/wagtailadmin/permissions/includes/collection_member_permissions_formset.html#L57-L71](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/permissions/includes/collection_member_permissions_formset.html#L57-L71)

1. wagtail/admin/templates/wagtailadmin/reports/

    - [ ] [wagtail/admin/templates/wagtailadmin/reports/workflow.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/reports/workflow.html#L83-L94)

1. wagtail/admin/templates/wagtailadmin/shared/

    - [ ] [wagtail/admin/templates/wagtailadmin/shared/locale_selector.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/shared/locale_selector.html#L5)
        NOTE: need to inspect if this breaks CSP

1. wagtail/admin/templates/wagtailadmin/widgets/

    - [ ] [wagtail/admin/templates/wagtailadmin/widgets/auto_height_text_input.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/widgets/auto_height_text_input.html#L1)
    - [ ] [wagtail/admin/templates/wagtailadmin/widgets/date_input.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/widgets/date_input.html#L1)
    - [ ] [wagtail/admin/templates/wagtailadmin/widgets/datetime_input.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/widgets/datetime_input.html#L1)
    - [ ] [wagtail/admin/templates/wagtailadmin/widgets/draftail_rich_text_area.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/widgets/draftail_rich_text_area.html#L1)
    - [ ] [wagtail/admin/templates/wagtailadmin/widgets/hallo_rich_text_area.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/widgets/hallo_rich_text_area.html#L1)
    - [ ] [wagtail/admin/templates/wagtailadmin/widgets/tag_widget.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/widgets/tag_widget.html#L2-L8)
    - [ ] [wagtail/admin/templates/wagtailadmin/widgets/time_input.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/widgets/time_input.html#L1)

1. wagtail/admin/templates/wagtailadmin/workflows/

    - [ ] [index.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/workflows/index.html#L105-L116)
    - [ ] [includes/_edit_js.html](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/workflows/includes/_edit_js.html#L1-L25)
    - [ ] [includes/workflow_pages_formset.html#L42-L48](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/workflows/includes/workflow_pages_formset.html#L42-L48)
    - [ ] [includes/workflow_pages_formset.html#L55-L69](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/admin/templates/wagtailadmin/workflows/includes/workflow_pages_formset.html#L55-L69)

### Wagtail Contrib

- [ ] [wagtail/contrib/forms/templates/wagtailforms/index_submissions.html#L7-L81](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/contrib/forms/templates/wagtailforms/index_submissions.html#L7-L81)
- [ ] [wagtail/contrib/modeladmin/templates/modeladmin/prepopulated_slugs.html#L2-L6](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/contrib/modeladmin/templates/modeladmin/prepopulated_slugs.html#L2-L6)
- [ ] [https://wagtail/contrib/modeladmin/templates/modeladmin/includes/search_form.html#L19](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/contrib/modeladmin/templates/modeladmin/includes/search_form.html#L19)
- [ ] [wagtail/contrib/redirects/templates/wagtailredirects/choose_import_file.html#L7-L13](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/contrib/redirects/templates/wagtailredirects/choose_import_file.html#L7-L13)
- [ ] [wagtail/contrib/redirects/templates/wagtailredirects/index.html#L9-L15](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/contrib/redirects/templates/wagtailredirects/index.html#L9-L15)
- [ ] [wagtail/contrib/search_promotions/templates/wagtailsearchpromotions/add.html#L45-L52](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/contrib/search_promotions/templates/wagtailsearchpromotions/add.html#L45-L52)
- [ ] [wagtail/contrib/search_promotions/templates/wagtailsearchpromotions/edit.html#L37-L44](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/contrib/search_promotions/templates/wagtailsearchpromotions/edit.html#L37-L44)
- [ ] [wagtail/contrib/search_promotions/templates/wagtailsearchpromotions/index.html#L7-L13](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/contrib/search_promotions/templates/wagtailsearchpromotions/index.html#L7-L13)
- [ ] [wagtail/contrib/search_promotions/templates/wagtailsearchpromotions/includes/searchpromotions_formset.html#L9-L13](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/contrib/search_promotions/templates/wagtailsearchpromotions/includes/searchpromotions_formset.html#L9-L13)
- [ ] [wagtail/contrib/styleguide/templates/wagtailstyleguide/base.html#L913-L926](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/contrib/styleguide/templates/wagtailstyleguide/base.html#L913-L926)

### Wagtail Documents

- [ ] [wagtail/documents/templates/wagtaildocs/documents/add.html#L12-L18](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/documents/templates/wagtaildocs/documents/add.html#L12-L18)
- [ ] [wagtail/documents/templates/wagtaildocs/documents/edit.html#L12-L18](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/documents/templates/wagtaildocs/documents/edit.html#L12-L18)
- [ ] [wagtail/documents/templates/wagtaildocs/documents/index.html#L6-L18](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/documents/templates/wagtaildocs/documents/index.html#L6-L18)
- [ ] [wagtail/documents/templates/wagtaildocs/multiple/add.html#L60-L76](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/documents/templates/wagtaildocs/multiple/add.html#L60-L76)
- [ ] [wagtail/documents/templates/wagtaildocs/multiple/add.html#L94-L101](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/documents/templates/wagtaildocs/multiple/add.html#L94-L101)

### Wagtail Images

- [ ] [wagtail/images/templates/wagtailimages/images/add.html#L12-L18](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/images/templates/wagtailimages/images/add.html#L12-L18)
- [ ] [wagtail/images/templates/wagtailimages/images/edit.html#L20-L26](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/images/templates/wagtailimages/images/edit.html#L20-L26)
- [ ] [wagtail/images/templates/wagtailimages/images/index.html#L9-L21](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/images/templates/wagtailimages/images/index.html#L9-L21)
- [ ] [wagtail/images/templates/wagtailimages/multiple/add.html#L60-L82](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/images/templates/wagtailimages/multiple/add.html#L60-L82)
- [ ] [wagtail/images/templates/wagtailimages/multiple/add.html#L104-L117](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/images/templates/wagtailimages/multiple/add.html#L104-L117)

### Wagtail Snippets

- [ ] [wagtail/snippets/templates/wagtailsnippets/snippets/type_index.html#L7-L13](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/snippets/templates/wagtailsnippets/snippets/type_index.html#L7-L13)

### Wagtail Users

- [ ] [wagtail/users/templates/wagtailusers/groups/index.html#L6-L12](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/users/templates/wagtailusers/groups/index.html#L6-L12)
- [ ] [wagtail/users/templates/wagtailusers/groups/includes/page_permissions_formset.html#L45-L51](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/users/templates/wagtailusers/groups/includes/page_permissions_formset.html#L45-L51)
- [ ] [wagtail/users/templates/wagtailusers/users/index.html#L6-L16](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/users/templates/wagtailusers/users/index.html#L6-L16)

### Wagtail Utils

- [ ] [wagtail/utils/widgets.py#L30](https://github.com/wagtail/wagtail/blob/571b9e1918861bcbe64a9005a4dc65f3a1fe7a15/wagtail/utils/widgets.py#L30)
NOTE: this uses a unique approach in all of the wagtail codebase to inject inline scripts

### Wagtail Embeds

- [ ] [wagtail/embeds/templates/wagtailembeds/embed_frontend.html#L1](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/embeds/templates/wagtailembeds/embed_frontend.html#L1)

### Wagtail Admin

- [ ] [wagtail/admin/templates/wagtailadmin/home/upgrade_notification.html#L3](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/admin/templates/wagtailadmin/home/upgrade_notification.html#L3)
- [ ] [wagtail/admin/templates/wagtailadmin/notifications/base.html#L7-L82](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/admin/templates/wagtailadmin/notifications/base.html#L7-L82)
- [ ] [wagtail/admin/templates/wagtailadmin/notifications/base.html#L84](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/admin/templates/wagtailadmin/notifications/base.html#L84)
- [ ] [wagtail/admin/templates/wagtailadmin/notifications/base.html#L89](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/admin/templates/wagtailadmin/notifications/base.html#L89)
- [ ] [wagtail/admin/templates/wagtailadmin/notifications/base.html#L95](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/admin/templates/wagtailadmin/notifications/base.html#L95)
- [ ] [wagtail/admin/templates/wagtailadmin/notifications/base.html#L105](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/admin/templates/wagtailadmin/notifications/base.html#L105)
- [ ] [wagtail/admin/templates/wagtailadmin/notifications/base.html#L107](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/admin/templates/wagtailadmin/notifications/base.html#L107)
- [ ] [wagtail/admin/templates/wagtailadmin/notifications/base.html#L117](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/admin/templates/wagtailadmin/notifications/base.html#L117)
- [ ] [wagtail/admin/templates/wagtailadmin/pages/add_subpage.html#L23](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/admin/templates/wagtailadmin/pages/add_subpage.html#L23)
- [ ] [wagtail/admin/templates/wagtailadmin/pages/confirm_delete.html#L31](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/admin/templates/wagtailadmin/pages/confirm_delete.html#L31)
- [ ] [wagtail/admin/templates/wagtailadmin/pages/edit_alias.html#L5](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/admin/templates/wagtailadmin/pages/edit_alias.html#L5)
- [ ] [wagtail/admin/templates/wagtailadmin/pages/search_results.html#L19](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/admin/templates/wagtailadmin/pages/search_results.html#L19)
- [ ] [wagtail/admin/templates/wagtailadmin/pages/search_results.html#L26](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/admin/templates/wagtailadmin/pages/search_results.html#L26)
- [x] [wagtail/admin/templates/wagtailadmin/pages/workflow_history/index.html#L11](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/admin/templates/wagtailadmin/pages/workflow_history/index.html#L11)
- [ ] [wagtail/admin/templates/wagtailadmin/permissions/includes/collection_member_permissions_formset.html#L31](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/admin/templates/wagtailadmin/permissions/includes/collection_member_permissions_formset.html#L31)
- [ ] [wagtail/admin/templates/wagtailadmin/shared/icons.html#L2](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/admin/templates/wagtailadmin/shared/icons.html#L2)
- [ ] [wagtail/admin/templates/wagtailadmin/shared/locale_selector.html#L4](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/admin/templates/wagtailadmin/shared/locale_selector.html#L4)
- [ ] [wagtail/admin/templates/wagtailadmin/userbar/base.html#L9](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/admin/templates/wagtailadmin/userbar/base.html#L9)
- [ ] [wagtail/admin/templates/wagtailadmin/workflows/includes/workflow_pages_formset.html#L28](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/admin/templates/wagtailadmin/workflows/includes/workflow_pages_formset.html#L28)

### Contrib

- [ ] [wagtail/contrib/forms/templates/wagtailforms/list_submissions.html#L10](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/contrib/forms/templates/wagtailforms/list_submissions.html#L10)
- [ ] [wagtail/contrib/modeladmin/templates/modeladmin/index.html#L65](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/contrib/modeladmin/templates/modeladmin/index.html#L65)
- [ ] [wagtail/contrib/search_promotions/templates/wagtailsearchpromotions/includes/searchpromotion_form.html#L2](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/contrib/search_promotions/templates/wagtailsearchpromotions/includes/searchpromotion_form.html#L2)
- [ ] [wagtail/contrib/styleguide/templates/wagtailstyleguide/base.html#L327](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/contrib/styleguide/templates/wagtailstyleguide/base.html#L327)
- [ ] [wagtail/contrib/styleguide/templates/wagtailstyleguide/base.html#L336](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/contrib/styleguide/templates/wagtailstyleguide/base.html#L336)
- [ ] [wagtail/contrib/styleguide/templates/wagtailstyleguide/base.html#L345](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/contrib/styleguide/templates/wagtailstyleguide/base.html#L345)
- [ ] [wagtail/contrib/styleguide/templates/wagtailstyleguide/base.html#L468-L474](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/contrib/styleguide/templates/wagtailstyleguide/base.html#L468-L474)
- [ ] [wagtail/contrib/styleguide/templates/wagtailstyleguide/base.html#L757](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/contrib/styleguide/templates/wagtailstyleguide/base.html#L757)
- [ ] [wagtail/contrib/styleguide/templates/wagtailstyleguide/base.html#L777](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/contrib/styleguide/templates/wagtailstyleguide/base.html#L777)
- [ ] [wagtail/contrib/styleguide/templates/wagtailstyleguide/base.html#L789-L803](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/contrib/styleguide/templates/wagtailstyleguide/base.html#L789-L803)

### Documents

- [ ] [wagtail/documents/templates/wagtaildocs/multiple/add.html#L54](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/documents/templates/wagtaildocs/multiple/add.html#L54)
- [ ] [wagtail/documents/templates/wagtaildocs/multiple/add.html#L65](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/documents/templates/wagtaildocs/multiple/add.html#L65)

### Embeds

- [ ] [wagtail/embeds/templates/wagtailembeds/embed_frontend.html#L1](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/embeds/templates/wagtailembeds/embed_frontend.html#L1)

### Images

- [ ] [wagtail/images/templates/wagtailimages/images/edit.html#L66](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/images/templates/wagtailimages/images/edit.html#L66)
- [ ] [wagtail/images/templates/wagtailimages/multiple/add.html#L54](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/images/templates/wagtailimages/multiple/add.html#L54)
- [ ] [wagtail/images/templates/wagtailimages/multiple/add.html#L66](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/images/templates/wagtailimages/multiple/add.html#L66)

### Users

- [ ] [wagtail/users/templates/wagtailusers/groups/includes/page_permissions_formset.html#L31](https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/users/templates/wagtailusers/groups/includes/page_permissions_formset.html#L31)

**Comment by gasman** [comment]
Don't know if this is on your radar already, but a lot of inline `<script>` tags won't necessarily be in templates - in particular, form widgets inheriting from [WidgetWithScript](https://github.com/wagtail/wagtail/blob/main/wagtail/utils/widgets.py) will contribute a lot of them.

**Comment by brownaa** [comment]
@gasman Thanks for the tip.

**Comment by brownaa** [comment]
`script-src` is done.

**Comment by brownaa** [comment]
`style-src` is done

**Comment by brownaa** [comment]
Example of how to add nonce to an inline `<script>` element.
https://github.com/brownaa/wagtail/commit/0d497580581399cb52c148dd17899550f7febb94

Example of how to add nonce to an inline `style` element.
https://github.com/brownaa/wagtail/commit/bc2738e540731fe47dcdc6273f6122a36e45eaa8

Based on my testing, it does not appear that we could add a nonce to a style attribute like
`<div style="display:none" nonce="r@nd0m">`

**Comment by thibaudcolas** [comment]
Reading up / based on our discussion in Slack, we’ll need to get rid of those `style=""` attributes in HTML/Python, and in JavaScript switch any usage of `setAttribute('style')` to adding individual properties on the `style` DOM attribute: https://stackoverflow.com/a/29089970/1798491.

I couldn’t see any `setAttribute('style')` in Wagtail’s JS code, but there could be in our dependencies.

For `style` attributes in HTML, I could see two based on the list above:

- Responsive embeds. https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/embeds/templates/wagtailembeds/embed_frontend.html#L1

Does some clever math. This would either require a `<style>` attribute with a nonce and adding styles via a unique id to each iframe… or I think we might be able to refactor this to use the more modern `aspect-ratio` CSS property.

- Image chooser renditions. https://github.com/wagtail/wagtail/blob/b34d48297e737b5ebbe2d7f3d3af3fa6b527487f/wagtail/images/templates/wagtailimages/images/edit.html#L66

This might be easier by setting the values via JS.

**Comment by rob101** [comment]
Why not solve this with a Middleware class? E.g. assuming you are using django-csp to generate a nonce, you could follow something like this with BS4 to add attributes to tags:

```
from __future__ import annotations
from typing import Callable
from bs4 import BeautifulSoup

from django.http import HttpRequest, HttpResponse


class ScriptMiddleware:
    def __init__(self,
                 get_response: Callable[[HttpRequest], HttpResponse]) -> None:
        self.get_response = get_response

    def __call__(self, request: HttpRequest) -> HttpResponse:
        response = self.get_response(request)
        content = response.content.decode(response.charset)

        # Add nonce to scripts
        soup = BeautifulSoup(content)
        els = soup.find_all('script')
        for el in els:
            el.attrs['nonce'] = str(request.csp_nonce)
        response.content = str(soup)

        return response
```

**Comment by gasman** [comment]
@rob101 Because that's defeating the point of having CSP in the first place. If someone can sneak a `<script>` tag into the template through an XSS attack, this middleware will blindly mark it as trusted even though there's no checking to see where it came from.

**Comment by rob101** [comment]
XSS is usually client side isn't it, typically a browser side script injection? The middleware runs before DOM load so bad actor script tags would not be caught, unless I am missing something?

**Comment by gasman** [comment]
@rob101 OK, yep, the third type described there (DOM Based XSS) is one that could conceivably be stopped by CSP with this middleware approach (in a setup where the malicious script isn't initially present in executable form as part of the served document, but is placed there as a result of running trusted client-side code with security flaws). I don't believe that's the most common type by a long way, though.

**Comment by lb-** [comment]
Workflow history detail will be solved via https://github.com/wagtail/wagtail/pull/8626

**Comment by lb-** [comment]
RFC 78 created https://github.com/wagtail/rfcs/pull/78  - with the goal of giving us a practical solution to widget scripts (amongst other things).

**Comment by lb-** [comment]
`wagtail/contrib/redirects/templates/wagtailredirects/choose_import_file.html` should be fixed by https://github.com/wagtail/wagtail/pull/8654

script was not even used for this view.

**Comment by lb-** [comment]
Login page - inline script to focus on username field has been replaced with autofocus attribute. 

Via #8925

**Comment by lb-** [comment]
#10045 will fix some of the `script` elements used as template tags in `documents|images/mulitiple/add`.

Hopefully will get a chance to update this main issue with all the things that are checked off.

**Comment by thibaudcolas** [comment]
I’ve updated `font-src` and `img-src` based on recent progress, testing by @robhudson, and #12368.

Rob’s CSP policy for reference:

```
frame-src 'self';
style-src 'unsafe-inline' 'self';
default-src 'self';
connect-src 'self';
script-src 'unsafe-eval' 'unsafe-inline' 'self';
font-src 'self';
child-src 'self';
img-src 'self' data:;
```

**Comment by lb-** [comment]
Some PRs recently merged, or still needing review, that get us closer to removing usage of inline scripts. 

- https://github.com/wagtail/wagtail/pull/11633
- https://github.com/wagtail/wagtail/pull/12360
- https://github.com/wagtail/wagtail/pull/12361
- https://github.com/wagtail/wagtail/pull/11166
- https://github.com/wagtail/wagtail/pull/11620
- https://github.com/wagtail/wagtail/pull/11775

**Comment by thibaudcolas** [comment]
As far as I can see, aside from #12368 for `img-src`, we’ll only need fixes for `style-src` and `script-src` from now on. There are few remaining, so I’ve taken the time to now split them across separate issues so it’s easier for contributors to take on, and will now close this issue – we’re getting there!

- #12937
- #12940
- #12938
- #12939
- https://github.com/springload/draftail/issues/460

## CSP testing

Here is the March 2025 "strictest simplest CSP" which I believe can be used for the Wagtail admin:

```txt
default-src 'self';
script-src 'self' 'unsafe-inline';
style-src 'self' 'unsafe-inline';
img-src 'self' data: blob:;
```

And here is a variation [compatible with the bakerydemo](https://github.com/wagtail/bakerydemo), which I would recommend contributors interested in CSP support make use of:

```txt
default-src 'self';
script-src 'self' 'report-sample';
style-src 'self' 'report-sample';
img-src 'self' blob: i.ytimg.com www.gravatar.com;
connect-src 'self' releases.wagtail.org;
frame-src 'self' www.youtube.com;
```

## CSP scanner

I’ve also set up a pretty basic [CSP scanner](https://github.com/thibaudcolas/wagtail-tooling/tree/main/csp) with browser automation, which I hope we could reuse to add QA checks for this in Wagtail at some point. Here is what that reported for [bakerydemo](https://github.com/wagtail/bakerydemo).

<details>

<summary>Scan results</summary>


### 2025-03-01 bakerydemo style-src

```csv
url,line
icons.js,0
vendor.js,1
/media/images/vik.original.jpg,0
/documents/multiple/add/,551
/images/43/,783
/images/43/generate_url/,18
/images/multiple/add/,557
/pages/1/add_subpage/,298
/pages/1/add_subpage/,315
/pages/1/add_subpage/,332
/pages/1/add_subpage/,349
/pages/1/add_subpage/,366
/pages/1/add_subpage/,383
/pages/1/add_subpage/,400
/pages/1/add_subpage/,417
/styleguide/,1776
/styleguide/,3158
/styleguide/,3282
/styleguide/,3290
/styleguide/,694
/styleguide/,703
/styleguide/,712
```

### 2025-03-01 bakerydemo script-src

```csv
file,line
vendor.js,1
/forms/,601
/forms/,603
/forms/submissions/69/,576
/forms/submissions/69/,578
/groups/edit/1/,1613
/groups/edit/2/,1617
/pages/,606
/pages/,608
/pages/3/edit/,1820
/pages/34/edit/,1856
/pages/34/edit/,2223
/pages/60/,699
/pages/60/,701
/pages/60/edit/,1895
/pages/60/edit/,2148
/pages/60/edit/,2326
/pages/60/edit/,2764
/pages/60/edit/,2964
/pages/60/edit/,3164
/pages/60/history/,682
/pages/60/history/,684
/pages/61/edit/,1820
/pages/62/edit/,1930
/pages/62/edit/,2089
/pages/62/edit/,2320
/pages/62/edit/,2554
/pages/63/edit/,1820
/pages/65/edit/,1855
/pages/65/edit/,2331
/pages/65/edit/,2388
/pages/65/edit/,2639
/pages/65/edit/,2696
/pages/65/edit/,2947
/pages/65/edit/,3004
/pages/65/edit/,3255
/pages/65/edit/,3312
/pages/65/edit/,3563
/pages/65/edit/,3620
/pages/65/edit/,3871
/pages/65/edit/,3928
/pages/65/edit/,4179
/pages/65/edit/,4236
/pages/65/edit/,4640
/pages/68/edit/,1970
/pages/68/edit/,2129
/pages/68/edit/,2360
/pages/68/edit/,2594
/pages/69/edit/,1800
/pages/69/edit/,4560
/pages/70/edit/,1898
/pages/72/edit/,1930
/pages/72/edit/,2089
/pages/72/edit/,2320
/pages/72/edit/,2554
/pages/81/edit/,1699
/pages/81/edit/,2329
/pages/81/edit/,2563
/pages/82/copy/,476
/pages/82/edit/,1860
/pages/82/edit/,2490
/pages/82/edit/,2724
/pages/82/history/,728
/pages/82/history/,730
/pages/82/move/,401
/pages/add/base/formpage/1/,1342
/pages/add/base/formpage/1/,1932
/pages/add/base/gallerypage/1/,1456
/pages/add/base/homepage/1/,1312
/pages/add/base/homepage/1/,1565
/pages/add/base/homepage/1/,1743
/pages/add/base/homepage/1/,2181
/pages/add/base/homepage/1/,2381
/pages/add/base/homepage/1/,2581
/pages/add/base/standardpage/1/,1456
/pages/add/blog/blogindexpage/1/,1371
/pages/usage/base/formpage/,615
/pages/usage/base/formpage/,617
/redirects/1/,751
/redirects/add/,713
/reports/aging-pages/,623
/reports/locked/,616
/reports/locked/,618
/reports/site-history/,720
/reports/site-history/,722
/reports/workflow_tasks/,816
/reports/workflow_tasks/,818
/reports/workflow_tasks/,865
/reports/workflow_tasks/,867
/reports/workflow/,805
/reports/workflow/,807
/searchpicks/1/,75
/searchpicks/1/,814
/searchpicks/add/,75
/searchpicks/reports/search-terms/,586
/searchpicks/reports/search-terms/,588
/sites/edit/2/,750
/sites/new/,705
/snippets/base/footertext/,654
/snippets/base/footertext/,656
/snippets/base/person/,810
/snippets/base/person/,812
/snippets/base/person/add/,1245
/snippets/base/person/edit/1/,1598
/snippets/base/person/history/1/,716
/snippets/base/person/history/1/,718
/snippets/breads/breadingredient/,646
/snippets/breads/breadingredient/,648
/snippets/breads/breadtype/,582
/snippets/breads/breadtype/,584
/styleguide/,2068
/styleguide/,2110
/styleguide/,2152
/styleguide/,2659
/styleguide/,2772
/styleguide/,2886
/styleguide/,3000
/users/,691
/users/,693
/workflows/add/,818
/workflows/add/,994
/workflows/edit/1/,1009
/workflows/edit/1/,1192
/workflows/edit/1/,1311
/workflows/edit/1/,779
/workflows/usage/1/,655
/workflows/usage/1/,657
```

</details>

**Comment by lb-** [comment]
Thank you @thibaudcolas - this is a great report, we've had some incredible momentum on this in the last few years. 

Thanks for splitting out the remaining tasks.
